### DISCUSSION TITLE
Adding UI tests for mocking paper actions using Jest/Enzyme

### DISCUSSION THREAD

**Discussion by Anuthama-Ganesan** [clientIO/joint#1913] [Q&A]
Hello,

I am trying to add unit tests for a rapid js based file and I am trying to mock the UI events in joint UI. Are there any specific framework and example tests you can provide for the same?
I am also attaching my source code of the component file here - 
https://codesandbox.io/s/hungry-nash-784w8b?file=/src/Genealogy.tsx
It would be great if there are any examples based on the events added over there - on hover, on click, etc. 
I am currently mocking the redux states used in my component and rendering them using "mount". Let me know if I have to provide any other detail.

**Comment by jamesgeorgewilliams** [comment]
You might be able to take advantage of `userEvent` in React `testing-library`
https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent

**Reply by Anuthama-Ganesan** [reply]
Is it possible to mount the joint paper component with enzyme framework and mimic the events? Using enzyme I am unable to figure out how I locate the paper components or individual shapes/icons to handle events.
Is there a way to deal with this kind of scenario? Any examples for this using Jest/enzyme?
Using react testing library is it possible to uniquely identify a cell when the whole paper is rendered, inorder to simulate these events?

**Reply by Anuthama-Ganesan** [reply]
Any possible workaround for this?

**Reply by jamesgeorgewilliams** [reply]
In general, it should be possible to query individual elements/cells with frameworks that provide this functionality. 

One simple method you can use is to add a custom class in the Cell `markup` and then query this class. We use this method in the following tutorial: https://resources.jointjs.com/tutorial/testing-e2e-playwright

```js 
const rect = new shapes.standard.Rectangle({
      position: { x: 200, y: 200 },
      size: { width: 100, height: 100 },
      attrs: {
          label: {
            text: 'Hello World'
          }
      },
      markup: [{
         tagName: 'rect',
         selector: 'body'
       }, {
         tagName: 'text',
         selector: 'label',
         className: 'custom-label-class'
      }]
});
```

I am not very familiar with `Enzyme`, but I imagine querying in this manner should not be difficult. `Enzyme` provides `find` for example which works with CSS selectors, and also some other ways in which to query.
https://enzymejs.github.io/enzyme/docs/api/ShallowWrapper/find.html
https://enzymejs.github.io/enzyme/docs/api/selector.html

It also provides a way to simulate events:
https://enzymejs.github.io/enzyme/docs/api/ShallowWrapper/simulate.html

Using React `testing-library` with similar ideas, you could do the following:

Our example event will change the label text on cell click:
```js
paper.on('cell:pointerclick', function(elementView) {  
     const currentElement = elementView.model;
     currentElement.attr('label/text', 'newText');
});
```

Our `App.test.tsx` file will be as follows:
```js
import { render } from '@testing-library/react';
import { within, waitFor } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';
import App from './App';

test('change element text onClick', async () => {

  const user = userEvent.setup();
  render(<App />);
  
  const label = document.querySelector('.custom-label-class');
  expect(label).toBeInTheDocument();
  within(label).getByText('Hello World');

  await user.click(label);

  await waitFor(() => {
    within(label).getByText('newText');
  });

});
```

It was also necessary for me to mock the following methods in `Jest` to perform this test with `testing-library`.

The `setupTests` file needed the following in addition to existing mocks.

```js

beforeEach(()=>{

  Object.defineProperty(global.document, 'elementFromPoint', {
    writable: true,
    value: jest.fn(),
  });

  Object.defineProperty(global.SVGElement.prototype, 'getScreenCTM', {
      writable: true,
      value: jest.fn(),
   });
  
});
```

This will successfully perform a test for `onClick`, but depending on what you are doing in your tests, you may need to mock some other missing functionality.

**Reply by Anuthama-Ganesan** [reply]
Thanks for the explanation, that should be really helpful.
I was trying to simulate a click event with enzyme and it is as follows -
```
it("simulate cell click", () => {
        const wrapper = mount(<Genealogy />);
        expect(wrapper).toBeTruthy();
        wrapper.find('.cell-data-wrap').at(0).simulate("click");
    });
```
I guess this should locate the cell and simulate a click, but the onClick event handler was not getting called as such. Not sure if I am missing anything. If there is, i'd be glad to know. 
On the other hand, let me give react-testing-library a try with the examples above and get back. Thanks.
